torch/jit.py

```diff
@@ -1,26 +1,496 @@
+        # output so we don't unflatten it when we get out
+        # NB: Not a method because trace_func_raw can't deal
+        # with methods
+        @_raw_trace(nderivs=self.nderivs)
+        def traced_inner(in_vars, in_struct):
+            return _flatten(self.inner(*args))
+
+        in_vars, in_struct = _flatten(args, self.state_dict(keep_vars=True).values())
+        trace, (out_vars, out_struct) = traced_inner(in_vars, in_struct)
+        out, extra = _unflatten(out_vars, out_struct)
+        assert len(extra) == 0
+        return trace, out
+
+
+# Functional version that assumes that all parameters are explicitly
+# specified
+def _raw_trace(nderivs=0):
+    def raw_trace(f):
+        # f takes two arguments, (in_vars, in_struct) (as determined
+        # by _flatten); furthermore, it must be the case that in_vars
+        # contains all Variable inputs (including parameters.)  It must
+        # produce two outputs, (out_vars, out_struct) (also as determined
+        # by _flatten).
+        @functools.wraps(f)
+        def wrapper(in_vars, in_struct=None):
+            trace = torch._C._tracer_enter(in_vars, nderivs)
+            out_vars, out_struct = f(in_vars, in_struct)
+            torch._C._tracer_exit(out_vars)
+            return trace, (out_vars, out_struct)
+        return wrapper
+    return raw_trace
+
+
+# Lifecycle of a compiler:
+#
+# - It is given an underlying function, which knows how to actually
+#   execute the code that we want to compile.
+# - When we encounter an input configuration for which we don't
+#   have an optimized trace, we run the underlying function, tracing its
+#   result.  The trace is not done yet, so we save it into our set of pending
+#   traces for that configuration.
+# - When we encounter an input configuration whose trace is "ready"
+#   (that is, we've seen all of the passes, so the trace contains
+#   forwards/backwards/etc), we compile it, and then register this
+#   as the compiled trace.
+# - When we encounter an input configuration whose trace is compiled,
+#   we just directly run the compiled trace.
 #
-# import torch.jit
-# model = model.RNNModel(args.model, ...)
-# model = torch.jit.traced(model)
+# You should never use this class directly; instead, use compile.  However,
+# the intended manual usage of this class looks like this:
+#
+#     class CompiledModel(_CompiledMixin, nn.Module):
+#         def forward(self, x):
+#             ...
+#     CompiledModule.init_compiler()
+#     model = CompiledModule()
+#
+class _CompiledMixin(object):
+    # Global over ALL compilations!  This helps us disambig if two Modules have
+    # the same __name__ but actually are different
+    __next_id = 0
+
+    @classmethod
+    def init_compiler(cls, name=None, enabled=True, time=False, **kwargs):
+        # Ensure we are not shadowing this method on the class we mixed with
+        assert not hasattr(super(_CompiledMixin, cls), "init_compiler")
```

Comment:
No you can't do it this way, because then the private members won't get mangled (mangling only occurs for functions syntactically inside the class definition.)